[clang][sema]: Error when SEH __try is used in a function with C++ dtors

Qwinci opened the pull request:
~~Don't generate calls to llvm.seh.scope.{begin,end} if c++ exceptions are disabled. Currently clang sometimes causes LLVM to crash due to invalid usage of the seh intrinsics when c++ objects with destructors are in scope with __try/__except when using -fasync-exceptions (see https://github.com/llvm/llvm-project/issues/82917). I am not sure why the assert that I removed fails with the changes, it would be nice if someone who knows more about the seh code would comment on this. CC @phoebewang~~
Add a sema check to make sure that there are no C++ objects with non-trivial destructors in a function where SEH __try is used.

phoebewang commented:
Is the problem happening when enabling async-exceptions but not cxx-exceptions on c++ objects? I'm not sure if it's a valid scenario. The designed requires `cxx-exceptions`, see descriptions in https://github.com/llvm/llvm-project/commit/797ad701522988e212495285dade8efac41a24d4

Maybe we can emit an error for such case instead?

Qwinci commented:
> Is the problem happening when enabling async-exceptions but not cxx-exceptions on c++ objects? I'm not sure if it's a valid scenario. The designed requires `cxx-exceptions`, see descriptions in [797ad70](https://github.com/llvm/llvm-project/commit/797ad701522988e212495285dade8efac41a24d4)
> 
> Maybe we can emit an error for such case instead?
Actually yeah my bad, looks like its invalid on MSVC to have objects with destructors in a function where __try/__except are used. Ill see if I can change this to be a check for that instead (and this is unrelated to exceptions being enabled, at least in my experience the C style seh __try/__except work just fine with async-exceptions + cxx exceptions disabled, I'd like to keep it that way as I use them in bare-metal C++ code where enabling c++ exceptions is undesirable plus async-exceptions should be usable in C code too).

Qwinci committed:
[clang][sema]: Error when SEH __try is used in a function with C++ dtors

SEH __try is not allowed in a function with C++ objects that have non-trivial
destructors.
